### DISCUSSION TITLE
Usage of "update" causing errors to be set on form state value properties

### DISCUSSION THREAD

**Discussion by jmora-fb** [react-hook-form/react-hook-form#6930] [Q&A]
Hi everyone, I have a CodeSandbox demonstrating two different approaches I've used to update objects created via `useFieldArray`. In this specific scenario, not all of the properties in the field array objects are exposed as fields in the UI (i.e., there isn't a 1:1 mapping between the number of UI fields and the number of properties in the object being changed). The sandbox presents two UI fields: a select dropdown & a text input. The change handler function for the dropdown is when the object updates are happening. This involves applying the selected dropdown value to the corresponding property as well as setting other properties on the object being updated.

As mentioned, on dropdown change, there are two approaches being used in the CodeSandbox:

Approach 1:
- Grab object via `getValues`
- Make a copy
- Change values directly on copied object
- Apply update via call to `update` from `useFieldArray`

Approach 2:
- Apply updates to individual properties on object via a set of `setValue` calls

Approach 1 is the one that is giving me problems and this is my primary reason for creating this discussion. The way the object is being retrieved & changed before the call to `update` may not be correct usage, so I'd like to get some feedback on that.

#### The problem being observed
In the set of reproduction steps that will follow, I am running across a scenario in which the _value_ corresponding to the dropdown is being set to the _validation error object_. Additionally, the _validation error property_ is being set to the selected _value_. Something seems to be getting mixed up when updating via the first approach.

#### To reproduce:
- Select a shape from dropdown
- Click "Submit"
- Note the correct `typeValue`and `colorInputError` values outputted in the console
- Select a different shape from the dropdown
    - Note that `typeValue` is initially set to the selected shape value, then `typeSelectError` is _also_ set to the _selected shape value_. Finally, `typeValue` ends up being `undefined` & `typeSelectError` ends up being `undefined`
- Click "Submit" again
- Note that `typeValue` is now set to the dropdown validation error itself -- not correct

These issues lead me to believe that the way the object is being updated and then passed to the `update` function is not the right usage so I would like some insight on what should be done instead for that approach. It's worth noting that if `clearErrors` is called right before the call to `update`, the outlined issues go away entirely, but this seems more like a "band-aid" fix on top of an underlying problem.

##### Accessing & logging of dropdown value:
<img width="418" alt="Screen Shot 2021-10-29 at 11 09 46 AM" src="https://user-images.githubusercontent.com/87672855/139485191-bb5781db-6e7d-4daf-a743-1ed2078c08ef.png">

##### Accessing & logging of validation errors:
<img width="622" alt="Screen Shot 2021-10-29 at 11 09 34 AM" src="https://user-images.githubusercontent.com/87672855/139485209-5479131c-ab1a-4c70-8f90-e490653d7dad.png">

![dropdown-sandbox](https://user-images.githubusercontent.com/87672855/139486566-7cd64096-9807-4988-a944-0adc498ac93f.gif)

----

The following CodeSandbox contains the code to reproduce the outlined scenario. The dropdown change handler (`onShapeTypeChange`) can be found within `ShapeField.tsx`. As mentioned, there are two approaches/variants. The `setValue` approach  marked `setValue variant` is commented out. The `update` approach marked `new object variant` is the one that is initially active. Note that the "new object variant" change handler that is initially active has a call to `clearErrors` commented out. Feel free to uncomment it to see how this "fixes" the issue.

#### [Link to CodeSandbox](https://codesandbox.io/s/react-hook-form-dropdown-value-update-forked-5zy6h)

For now, in the real application, the `setValue` approach/variant is being used since this does not cause any problems at all whatsoever and works exactly as expected. My intention behind this post is to try and figure out why the `update` approach is leading to these weird issues. Maybe manipulating the object & updating incorrectly, etc.

Thank you for the time and help. This post became somewhat lengthy, but it seems appropriate in order to provide sufficient context for better troubleshooting.

_Please excuse the typing errors on the field name strings. This is only a linting error and does not affect runtime functionality._

**Comment by bluebill1049** [comment]
thanks for the detailed question, can you reproduce this issue in simpler codesandobx.

>My intention behind this post is to try and figure out why the update approach is leading to these weird issues.

maybe focus on this API only.
